### PR TITLE
Feature/#2017 editor

### PR DESCRIPTION
**Pull request by jrobinAV**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
When locking a data node, providing an "editor_id" only prevents another editor from locking the data node; it does not prevent the other editor from writing the data node.

According to the documentation, it should prevent another editor from writing (appending, uploading) the data node.

## Related Tickets & Documents

- Related Issue #2017 
- Taking over PR #2122, abandoned by the contributor.

## Checklist

- [x] Does this solution meet the acceptance criteria of the related issue?
- [x] Is the related issue checklist completed?
- [x] Does this PR add unit tests for the developed code?
- [x] Was the documentation updated, or a dedicated issue for documentation created?

### CODE REVIEW

File: taipy/gui_core/_context.py [hunk 1]
```diff
@@ -1044,9 +1044,9 @@ def update_data(self, state: State, id: str, payload: t.Dict[str, str]):
                     else float(val)
                     if data.get("type") == "float"
                     else data.get("value"),
+                    editor_id=self.gui._get_client_id(),
```
Comment (jrobinAV) [inline taipy/gui_core/_context.py:1052]: @FredLL-Avaiga, I believe this answers the comment you made a few days ago on another PR.

Comment (FredLL-Avaiga) [reply taipy/gui_core/_context.py:1052]: it does indeed

Comment (FredLL-Avaiga) [reply taipy/gui_core/_context.py:1052]: we could use a public API  `get_state_id(state)` but `self.gui._get_client_id()` is already used when state is not available

File: taipy/gui_core/_context.py [hunk 2]
```diff
@@ -1028,7 +1028,7 @@ def update_data(self, state: State, id: str, payload: t.Dict[str, str]):
         if args is None or not isinstance(args, list) or len(args) < 1 or not isinstance(args[0], dict):
             return
         data = t.cast(dict, args[0])
-        error_var = payload.get("error_id")
+        error_var = data.get("error_id")
```
Comment (jrobinAV) [inline taipy/gui_core/_context.py:1036]: I changed that on Fred's advice.

File: taipy/gui_core/_context.py [hunk 3]
```diff
@@ -1217,18 +1219,19 @@ def on_dag_select(self, state: State, id: str, payload: t.Dict[str, str]):
 
     def on_file_action(self, state: State, id: str, payload: t.Dict[str, t.Any]):
         args = t.cast(list, payload.get("args"))
+        if args is None or not isinstance(args, list) or len(args) < 1 or not isinstance(args[0], dict):
+            return
```
Comment (jrobinAV) [inline taipy/gui_core/_context.py:1228]: I added this protection to be consistent with the `update_data` method.

File: tests/gui_core/test_context_tabular_data_edit.py [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        new_data.at[idx, col] = new_value
+        with patch("taipy.gui_core._context.is_readable", side_effect=is_true):
+            with patch("taipy.gui_core._context.is_editable", side_effect=is_true):
+                with patch("taipy.gui_core._context.core_get", side_effect=core_get) as mock_core_get:
+                    with patch.object(DataNode, "write") as mock_write:
+
+                        assign = self.__call_update_data(col, idx, new_value)
+
+                        mock_core_get.assert_called_once_with(dn.id)
+                        mock_write.assert_called_once()
```
Comment (jrobinAV) [inline tests/gui_core/test_context_tabular_data_edit.py:130]: I would expect the mock not to be called as the idx is incorrect. What do you think @FredLL-Avaiga ?

File: tests/gui_core/test_context_tabular_data_edit.py [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+        new_value = 100
+        new_data = data.copy()
+        new_data.at[idx, col] = new_value
+        with patch("taipy.gui_core._context.is_readable", side_effect=is_true):
+            with patch("taipy.gui_core._context.is_editable", side_effect=is_true):
+                with patch("taipy.gui_core._context.core_get", side_effect=core_get) as mock_core_get:
+                    with patch.object(DataNode, "write") as mock_write:
+                        assign = self.__call_update_data(col, idx, new_value)
+                        mock_core_get.assert_called_once_with(dn.id)
+                        mock_write.assert_called_once()
```
Comment (jrobinAV) [inline tests/gui_core/test_context_tabular_data_edit.py:153]: I would expect the mock not to be called as the column is incorrect. What do you think @FredLL-Avaiga ?

File: tests/gui_core/test_context_tabular_data_edit.py (hunk 4 above)
Comment (jrobinAV) [reply tests/gui_core/test_context_tabular_data_edit.py:130]: BTW that would make it consistent with the dict/list data format.

File: taipy/gui_core/_context.py [hunk 6]
```diff
@@ -1044,9 +1044,9 @@ def update_data(self, state: State, id: str, payload: t.Dict[str, str]):
                     else float(val)
                     if data.get("type") == "float"
                     else data.get("value"),
+                    editor_id=self.gui._get_client_id(),
                     comment=t.cast(dict, data.get(_GuiCoreContext.__PROP_ENTITY_COMMENT)),
                 )
-                entity.unlock_edit(self.gui._get_client_id())
```
Comment (FredLL-Avaiga) [inline taipy/gui_core/_context.py:1054]: I guess the unlock is done by write ?

File: taipy/gui_core/_context.py [hunk 7]
```diff
@@ -1217,18 +1219,19 @@ def on_dag_select(self, state: State, id: str, payload: t.Dict[str, str]):
...
         error_id = act_payload.get("error_id", "")
         if reason := is_readable(dn_id):
             try:
                 dn = t.cast(_FileDataNodeMixin, core_get(dn_id))
                 if act_payload.get("action") == "export":
-                    path = dn._get_downloadable_path()
-                    if path:
+                    if reason := dn.is_downloadable():
+                        path = dn._get_downloadable_path()
                         self.gui._download(Path(path), dn_id)
```
Comment (FredLL-Avaiga) [inline taipy/gui_core/_context.py:1238]: no need to check path if is_downloadable ?

File: tests/gui_core/test_context_tabular_data_edit.py (hunk 5 above)
Comment (FredLL-Avaiga) [reply tests/gui_core/test_context_tabular_data_edit.py:153]: how could the column be incorrect ?

File: tests/gui_core/test_context_tabular_data_edit.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply tests/gui_core/test_context_tabular_data_edit.py:130]: how could the idx be incorrect ?

File: taipy/gui_core/_context.py (hunk 6 above)
Comment (jrobinAV) [reply taipy/gui_core/_context.py:1054]: Yes, indeed. It was already the case, BTW.

File: taipy/gui_core/_context.py (hunk 7 above)
Comment (jrobinAV) [reply taipy/gui_core/_context.py:1238]: Absolutely. That's what `is_downloadable` does. It builds the Path, checks if it exists, and if it is a file.

File: tests/gui_core/test_context_tabular_data_edit.py (hunk 4 above)
Comment (jrobinAV) [reply tests/gui_core/test_context_tabular_data_edit.py:130]: We have a 3X3 table, and we want to update line 5.

File: tests/gui_core/test_context_tabular_data_edit.py (hunk 5 above)
Comment (jrobinAV) [reply tests/gui_core/test_context_tabular_data_edit.py:153]: We have a 2 column dataframe ("a" and "b"), and we want to update column "c".

Comment (FredLL-Avaiga) [reply tests/gui_core/test_context_tabular_data_edit.py:153]: that can't happen with the UI components

File: tests/gui_core/test_context_tabular_data_edit.py (hunk 4 above)
Comment (FredLL-Avaiga) [reply tests/gui_core/test_context_tabular_data_edit.py:130]: that can't happen with the UI component

Comment (jrobinAV) [reply tests/gui_core/test_context_tabular_data_edit.py:130]: OK so indeed, that's not a pb, thx.

File: taipy/core/data/_file_datanode_mixin.py [hunk 8]
```diff
@@ -65,8 +72,8 @@ def path(self) -> str:
     @path.setter
     def path(self, value) -> None:
         self._path = value
-        self.properties[self._PATH_KEY] = value
-        self.properties[self._IS_GENERATED_KEY] = False
+        self.properties[self._PATH_KEY] = value # type: ignore[attr-defined]
```
Comment (joaoandre-avaiga) [inline taipy/core/data/_file_datanode_mixin.py]: Needs rebasing after the changes in the pathing that were in #2267

**Pull request merged**